packages/core/src/new-api/internal/execution/execution-engine.ts

```diff
@@ -105,73 +131,487 @@ export class ExecutionEngine {
+      }
+      case "send-data": {
+        return this._convertRequestToSendDataTransaction(request, state);
+      }
+      case "deploy-contract":
+        return this._convertRequestToDeployTransaction(request, state);
+    }
+  }
+
+  private async _convertRequestToCallFunctionTransaction(
+    request: CallFunctionInteractionMessage,
+    state: ExecutionEngineState
+  ): Promise<ethers.providers.TransactionRequest> {
+    const artifact = await state.deploymentLoader.loadArtifact(
+      request.storedArtifactPath
+    );
+
+    const contractAddress: string = request.contractAddress;
+    const abi = artifact.abi;
+    const functionName: string = request.functionName;
+    const args: ArgumentType[] = request.args;
+    const value: bigint = BigInt(request.value);
+    const from: string = request.from;
+
+    const unsignedTx = state.chainDispatcher.constructCallTransaction(
+      contractAddress,
+      abi,
+      functionName,
+      args,
+      value,
+      from
+    );
+
+    return unsignedTx;
+  }
+
+  private async _convertRequestToSendDataTransaction(
+    request: SendDataInteractionMessage,
+    _state: ExecutionEngineState
+  ): Promise<ethers.providers.TransactionRequest> {
+    const unsignedTx: ethers.providers.TransactionRequest = {
+      from: request.from,
+      to: request.to,
+      value: BigInt(request.value),
+      data: request.data,
     };
+
+    return unsignedTx;
+  }
+
+  private async _convertRequestToDeployTransaction(
+    request: DeployContractInteractionMessage,
+    state: ExecutionEngineState
+  ): Promise<ethers.providers.TransactionRequest> {
+    const artifact = await state.deploymentLoader.loadArtifact(
+      request.storedArtifactPath
+    );
+
+    const abi = artifact.abi;
+    const args = request.args;
+    const value = BigInt(request.value);
+    const from = request.from;
+
+    // TODO: fix libraries
+    const linkedByteCode = await collectLibrariesAndLink(artifact, {});
+
+    const tx = state.chainDispatcher.constructDeployTransaction(
+      linkedByteCode,
+      abi,
+      args,
+      value,
+      from
+    );
+
+    return tx;
+  }
+
+  private async _readEventArg(
+    request: ReadEventArgumentInteractionMessage,
+    state: ExecutionEngineState
+  ): Promise<OnchainReadEventArgumentSuccessMessage | OnchainFailureMessage> {
+    const artifact = await state.deploymentLoader.loadArtifact(
+      request.storedArtifactPath
+    );
+
+    try {
+      const result = await state.chainDispatcher.getEventArgument(
+        request.eventName,
+        request.argumentName,
+        request.txToReadFrom,
+        request.eventIndex,
+        request.emitterAddress,
+        artifact.abi
+      );
+
+      return {
+        type: "onchain-result",
+        subtype: "read-event-arg-success",
+        futureId: request.futureId,
+        executionId: request.executionId,
+        result,
+      };
+    } catch (error) {
+      return {
```

Comment:
I am not sure about having the transaction handling here. We are letting ethers knowledge escape from the chain dispatcher.

Similarly we need to refine the error capture to be a more specific error value and code that Ignition know.

Reply:
This will cleaned up with error handling.